spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrResultSet.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2017-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.vision;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.TreeMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import com.google.cloud.storage.Blob;
+import com.google.cloud.vision.v1.AnnotateFileResponse;
+import com.google.cloud.vision.v1.AnnotateImageResponse;
+import com.google.cloud.vision.v1.TextAnnotation;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.util.JsonFormat;
+
+/**
+ * Represents the parsed OCR content for an document in the provided range of pages.
+ *
+ * @author Daniel Zou
+ */
+public class DocumentOcrResultSet {
+
+	private static final Pattern OUTPUT_PAGE_PATTERN = Pattern.compile("output-(\\d+)-to-(\\d+)\\.json");
+
+	private final TreeMap<Integer, OcrPageRange> ocrPageRanges;
+
+	private final TreeMap<Integer, List<TextAnnotation>> documentPagesCache;
```

Comment:
Can we instead just use one simple `HashMap<Integer, OcrPageRange>`, where there is an entry for every single page? The `getTextAnnotations()` can move to the `OcrPageRange` class, and caching can very easily be done there. I think it would simplify your code quiet a bit.

Reply:
Partially done. I moved the caching into OcrPageRange which simplified the caching code.

However, I didn't switch from TreeMap to HashMap. I don't think using TreeMap here is that much more complex that HashMap, so might as well do things efficiently. For users processing with JSON `batchSize = 100` we would reduce the number of entries stored in map by factor of 100.